dora/src/typeck/tests.rs

```diff
@@ -70,22 +70,36 @@ fn type_class_method_call() {
 }
 
 #[test]
-fn type_module_method_call_1() {
+fn type_module_method_call() {
     ok("module Foo {
                 fun bar() {}
                 fun baz() -> Int { return 1; }
             }
 
-            fun f1() { Foo.bar(); }
+            class Foo() {}
+
             fun f2() { Foo::bar(); }
-            fun g() -> Int { return Foo.baz(); }");
+            fun g() -> Int { return Foo::baz(); }");
+
+    err(
+        "module Foo {
+                 fun foo() -> Bar = Bar();
+               }
+               class Bar[T]
+
+               fun bar() {
+                 //Foo::foo(); // panicked at 'out-of-bounds', dora/src/ty.rs:732:33
```

Comment:
Stacktrace is
```
  13: <dora::ty::TypeList as core::ops::index::Index<usize>>::index
             at dora/src/ty.rs:732
  14: dora::semck::specialize::replace_type_param
             at dora/src/semck/specialize.rs:394
  15: dora::semck::specialize::replace_type_param::{{closure}}
             at dora/src/semck/specialize.rs:403
  16: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/212b2c7da87f3086af535b33a9ca6b5242f2d5a7/src/libcore/ops/function.rs:285
  17: core::option::Option<T>::map
             at /rustc/212b2c7da87f3086af535b33a9ca6b5242f2d5a7/src/libcore/option.rs:450
  18: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/212b2c7da87f3086af535b33a9ca6b5242f2d5a7/src/libcore/iter/adapters/mod.rs:791
  19: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
             at /rustc/212b2c7da87f3086af535b33a9ca6b5242f2d5a7/src/liballoc/vec.rs:2007
  20: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/212b2c7da87f3086af535b33a9ca6b5242f2d5a7/src/liballoc/vec.rs:1919
  21: core::iter::traits::iterator::Iterator::collect
             at /rustc/212b2c7da87f3086af535b33a9ca6b5242f2d5a7/src/libcore/iter/traits/iterator.rs:1558
  22: dora::semck::specialize::replace_type_param
             at dora/src/semck/specialize.rs:401
  23: dora::typeck::lookup::MethodLookup::find
             at dora/src/typeck/lookup.rs:296
  24: dora::typeck::expr::TypeCheck::check_expr_call_method
             at dora/src/typeck/expr.rs:1259
  25: dora::typeck::expr::TypeCheck::check_expr_call
             at dora/src/typeck/expr.rs:990
  26: dora::typeck::expr::TypeCheck::check_expr
             at dora/src/typeck/expr.rs:2041
  27: <dora::typeck::expr::TypeCheck as dora_parser::ast::visit::Visitor>::visit_stmt
             at dora/src/typeck/expr.rs:2074
  28: dora::typeck::expr::TypeCheck::check
             at dora/src/typeck/expr.rs:41
  29: dora::typeck::check
             at dora/src/typeck.rs:36
  30: dora::semck::check
             at dora/src/semck.rs:111
  31: dora::test::parse_with_errors
             at dora/src/test.rs:44
  32: dora::semck::tests::err
             at dora/src/semck.rs:539
  33: dora::typeck::tests::type_module_method_call
             at dora/src/typeck/tests.rs:84
  34: dora::typeck::tests::type_module_method_call::{{closure}}
             at dora/src/typeck/tests.rs:73
```

Reply:
Hmm, the program should be invalid since `foo` returns `Bar`, however `Bar` should have a type argument. It seems this is not checked in modules however.

Reply:
I updated the test, the `WrongNumberTypeParams` is reported correctly.

In the updated test, the generics are used correctly, but it still fails.

Reply:
See below!